tools/om-linker/om-linker.cpp

```diff
@@ -0,0 +1,162 @@
+using namespace mlir;
+using namespace circt;
+
+static cl::OptionCategory mainCategory("om-linker Options");
+
+static cl::list<std::string> inputFilenames(cl::Positional, cl::OneOrMore,
+                                            cl::desc("<input files>"),
+                                            cl::cat(mainCategory));
+
+static cl::opt<std::string> outputFilename("o", cl::desc("Output filename"),
+                                           cl::value_desc("filename"),
+                                           cl::init("-"),
+                                           cl::cat(mainCategory));
+
+/// This implements the top-level logic for the om-linker command, invoked once
+/// command line options are parsed and LLVM/MLIR are all set up and ready to
+/// go.
+static LogicalResult executeOMLinker(MLIRContext &context) {
+  // Create the timing manager we use to sample execution times.
+  DefaultTimingManager tm;
+  applyDefaultTimingManagerCLOptions(tm);
+  auto ts = tm.getRootScope();
+
+  auto numFiles = inputFilenames.size();
+
+  // Set up the input files.
+  struct ParsedInput {
+    StringRef name;
+    OwningOpRef<ModuleOp> mod;
+    SourceMgr mgr;
+  };
+  SmallVector<ParsedInput> srcs;
+  srcs.resize(numFiles);
+  auto parserTimer = ts.nest("Parsing inputs");
+  auto loadFile = [&](size_t i) -> LogicalResult {
+    auto &s = srcs[i];
+    s.name = inputFilenames[i];
+    auto fileParseTimer = parserTimer.nest(s.name);
+    s.mod = parseSourceFile<ModuleOp>(s.name, s.mgr, &context);
+    if (!s.mod)
+      return failure();
+    return success();
+  };
+
+  if (failed(failableParallelForEachN(&context, 0, numFiles, loadFile))) {
+    errs() << "error reading inputs\n";
+    return failure();
+  }
+
+  // This is the result module we are linking into.
+  OwningOpRef<ModuleOp> module(ModuleOp::create(UnknownLoc::get(&context)));
+
+  // Concat input modules without flatting ModuleOp. For example we construct
+  // an IR like this:
+  // ```
+  // module {
+  //   module {
+  //     om.class @A(%arg: i1) {}
+  //   }
+  //   module {
+  //     om.class.extern @A(%arg: i1) {}
+  //     om.class @B(%arg: i2) {}
+  //   }
+  // }
+  // ```
+  // Actual linking will be performed by the subsequent pass pipeline.
+  auto builder = OpBuilder::atBlockEnd(module.get().getBody());
+  for (auto &in : srcs)
+    builder.insert(in.mod.get());
+
+  // TODO: Apply a linking pass here.
+
+  // Create the output file.
+  std::string errorMessage;
+  auto outputFile = openOutputFile(outputFilename, &errorMessage);
+  if (!outputFile) {
+    errs() << errorMessage << "\n";
+    return failure();
+  }
+
+  // Dump output.
+  module->print(outputFile->os());
+
+  // If the result succeeded and we're emitting a file, close it.
+  outputFile->keep();
+
+  // We intentionally "leak" the Module into the MLIRContext instead of
+  // deallocating it.  There is no need to deallocate it right before process
+  // exit.
+  for (auto &src : srcs)
```

Comment:
nit: Consider doing the sort of thing I did in my `firld` thing (store in struct that leaks them in destructor), so that they're leaked on exit paths too not just success.

Reply:
It's arguably not worth making things gross for making error case faster, but just a thought :+1: .

Reply:
Done, thank you for the suggestion :)